src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -438,6 +443,7 @@ private BulkResponse execute() throws Exception {
             batch.size(),
             e
         );
+        recordsToReportOnError.keySet().removeAll(batch);
```

Comment:
To keep the symmetry with the null check in `put` I'd suggest adding another helper method in these calls.
Example: 

```
 private void removeReportedRecords(List<R> batch) {
    if (reporter != null) {
      recordsToReportOnError.keySet().removeAll(batch);
    }
  }
```